polymer/src/amp-my-embed.html

```diff
@@ -0,0 +1,285 @@
+        this.shadowAmp_ = null;
+
+        /**
+         * The root node of the shadow AMP.
+         * @note A single node must not be reused for multiple shadow AMP docs.
+         * @type {Element}
+         */
+        this.shadowRoot_ = null;
+
+        /** @private */
+        this.boundClickListener_ = this.clickListener_.bind(this);
+
+        /**
+         * Child element that will wrap the AMP shadow root.
+         * @private
+         * @type {Element}
+         */
+        this.container_ = document.getElementById('container');
+
+      }
+
+      ready() {
+        super.ready();
+        this.$.name.textContent = this.shadowAmp_;
+      }
+
+  /**
+   * Fetches the AMP document at `url` and attaches it as a shadow root.
+   * @private
+   * @param {string} url
+   */
+  fetchAndAttachAmpDoc_(url) {
+    this.fetchDocument_(url).then(doc => {
+      return this.ampReadyPromise_.then(amp => {
+        // Hide navigational and other unwanted elements before displaying.
+        this.hideUnwantedElementsOnDocument_(doc);
+
+        // Replace the old shadow root with a new div element.
+        const oldShadowRoot = this.shadowRoot_;
+        this.shadowRoot_ = document.createElement('div');
+        if (oldShadowRoot) {
+          this.$.name.replaceChild(this.shadowRoot_, oldShadowRoot);
+        } else {
+          this.$.name.appendChild(this.shadowRoot_);
+        }
+
+        // Attach the shadow document to the new shadow root.
+        this.shadowAmp_ = amp.attachShadowDoc(this.shadowRoot_, doc, url);
+      });
+    }).catch(error => {
+      this.setState({'offline': true});
+    });
+  }
+
+  /**
+   * Cleans up internal state of current shadow AMP document.
+   * @private
+   */
+  closeShadowAmpDoc_() {
+    if (typeof this.shadowAmp_.close === 'function') {
+      this.shadowAmp_.close();
+    }
+  }
+
+  /**
+   * Hides elements (e.g. banners) that would clash with the app shell.
+   * @param {!Document} doc
+   * @private
+   */
+   hideUnwantedElementsOnDocument_(doc) {
+     const banners = doc.getElementsByClassName('banner');
+     for (let i = 0; i < banners.length; i++) {
+       banners[i].style.display = 'none';
+     }
+   }
+
+  /**
+   * Fetches and parses HTML at `url`.
+   * @private
+   * @param {string} url
+   * @return {!Promise<!Document|!string>} If fetch succeeds, resolved with {!Document}.
+   *         Otherwise, rejects with {!string} error description.
+   */
+  fetchDocument_(url) {
+    return new Promise((resolve, reject) => {
+      this.xhr_ = new XMLHttpRequest();
+      this.xhr_.open('GET', url, true);
+      this.xhr_.responseType = 'document';
+      // This is set to text/* instead of text/html because the development server
+      // only forwards requests to the proxy for requests whose 'Accept' header
+      // is NOT text/html.
+      // https://github.com/facebookincubator/create-react-app/blob/master/packages/react-scripts/template/README.md#proxying-api-requests-in-development
```

Comment:
I assume this is copy-pasted from the React version? This function is very hard to follow.

Reply:
Removed those comments thanks!
The fetchDocument function comes from the Polymer News sample: https://github.com/Polymer/news/blob/amp/src/amp-viewer.html#L97. Is that the bit very hard to follow?